File: python/repository.py
import json
from datetime import datetime, timezone

from sqlalchemy import text
from sqlalchemy.orm import Session


class KillmailRepository:
    def __init__(self, session: Session):
        self.session = session

    # ── Write (poller) ──────────────────────────────────────────────

    def save(self, killmail: dict) -> bool:
        killmail_id = killmail["killmail_id"]
        esi = killmail["esi"]
        zkb = killmail["zkb"]
        victim = esi["victim"]
        position = victim.get("position") or {}

        result = self.session.execute(
            text("""
                INSERT IGNORE INTO killmails (
                    killmail_id, hash, killmail_time, solar_system_id, sequence_id,
                    war_id, moon_id,
                    victim_character_id, victim_corporation_id, victim_alliance_id,
                    victim_faction_id, victim_ship_type_id, victim_damage_taken,
                    victim_pos_x, victim_pos_y, victim_pos_z,
                    zkb_location_id, zkb_fitted_value, zkb_dropped_value,
                    zkb_destroyed_value, zkb_total_value, zkb_points,
                    zkb_is_npc, zkb_is_solo, zkb_is_awox,
                    zkb_labels, zkb_href, zkb_attacker_count, uploaded_at
                ) VALUES (
                    :killmail_id, :hash, :killmail_time, :solar_system_id, :sequence_id,
                    :war_id, :moon_id,
                    :victim_character_id, :victim_corporation_id, :victim_alliance_id,
                    :victim_faction_id, :victim_ship_type_id, :victim_damage_taken,
                    :victim_pos_x, :victim_pos_y, :victim_pos_z,
                    :zkb_location_id, :zkb_fitted_value, :zkb_dropped_value,
                    :zkb_destroyed_value, :zkb_total_value, :zkb_points,
                    :zkb_is_npc, :zkb_is_solo, :zkb_is_awox,
                    :zkb_labels, :zkb_href, :zkb_attacker_count, :uploaded_at
                )
            """),
            {
                "killmail_id": killmail_id,
                "hash": killmail["hash"],
                "killmail_time": esi["killmail_time"],
                "solar_system_id": esi["solar_system_id"],
                "sequence_id": killmail["sequence_id"],
                "war_id": esi.get("war_id"),
                "moon_id": esi.get("moon_id"),
                "victim_character_id": victim.get("character_id"),
                "victim_corporation_id": victim.get("corporation_id"),
                "victim_alliance_id": victim.get("alliance_id"),
                "victim_faction_id": victim.get("faction_id"),
                "victim_ship_type_id": victim["ship_type_id"],
                "victim_damage_taken": victim["damage_taken"],
                "victim_pos_x": position.get("x"),
                "victim_pos_y": position.get("y"),
                "victim_pos_z": position.get("z"),
                "zkb_location_id": zkb.get("locationID"),
                "zkb_fitted_value": zkb.get("fittedValue", 0),
                "zkb_dropped_value": zkb.get("droppedValue", 0),
                "zkb_destroyed_value": zkb.get("destroyedValue", 0),
                "zkb_total_value": zkb.get("totalValue", 0),
                "zkb_points": zkb.get("points", 0),
                "zkb_is_npc": int(zkb.get("npc", False)),
                "zkb_is_solo": int(zkb.get("solo", False)),
                "zkb_is_awox": int(zkb.get("awox", False)),
                "zkb_labels": json.dumps(zkb.get("labels", [])),
                "zkb_href": zkb.get("href"),
                "zkb_attacker_count": len(esi.get("attackers", [])),
                "uploaded_at": datetime.fromtimestamp(
                    killmail["uploaded_at"], tz=timezone.utc
                ).strftime("%Y-%m-%d %H:%M:%S"),
            },
        )

        if result.rowcount == 0:
            self.session.rollback()
            return False

        self._insert_attackers(killmail_id, esi.get("attackers", []))
        self._insert_items(killmail_id, victim.get("items", []))
        self.session.commit()
        return True

    def _insert_attackers(self, killmail_id: int, attackers: list[dict]) -> None:
        if not attackers:
            return
        for attacker in attackers:
            self.session.execute(
                text("""
                    INSERT INTO killmail_attackers (
                        killmail_id, character_id, corporation_id, alliance_id, faction_id,
                        ship_type_id, weapon_type_id, damage_done, final_blow, security_status
                    ) VALUES (
                        :killmail_id, :character_id, :corporation_id, :alliance_id, :faction_id,
                        :ship_type_id, :weapon_type_id, :damage_done, :final_blow, :security_status
                    )
                """),
                {
                    "killmail_id": killmail_id,
                    "character_id": attacker.get("character_id"),
                    "corporation_id": attacker.get("corporation_id"),
                    "alliance_id": attacker.get("alliance_id"),
                    "faction_id": attacker.get("faction_id"),
                    "ship_type_id": attacker.get("ship_type_id"),
                    "weapon_type_id": attacker.get("weapon_type_id"),
                    "damage_done": attacker.get("damage_done", 0),
                    "final_blow": int(attacker.get("final_blow", False)),
                    "security_status": attacker.get("security_status", 0),
                },
            )

    def _insert_items(
        self, killmail_id: int, items: list[dict], parent_id: int | None = None
    ) -> None:
        if not items:
            return
        for item in items:
            result = self.session.execute(
                text("""
                    INSERT INTO killmail_items (
                        killmail_id, parent_id, item_type_id, flag,
                        quantity_destroyed, quantity_dropped, singleton
                    ) VALUES (
                        :killmail_id, :parent_id, :item_type_id, :flag,
                        :quantity_destroyed, :quantity_dropped, :singleton
                    )
                """),
                {
                    "killmail_id": killmail_id,
                    "parent_id": parent_id,
                    "item_type_id": item["item_type_id"],
                    "flag": item.get("flag", 0),
                    "quantity_destroyed": item.get("quantity_destroyed", 0),
                    "quantity_dropped": item.get("quantity_dropped", 0),
                    "singleton": item.get("singleton", 0),
                },
            )
            if item.get("items"):
                inserted_id = result.lastrowid
                self._insert_items(killmail_id, item["items"], inserted_id)

    # ── Read (API) ──────────────────────────────────────────────────

    def list_kills(
        self,
        *,
        limit: int = 50,
        offset: int = 0,
        min_value: float | None = None,
        max_value: float | None = None,
        solar_system_id: int | None = None,
        ship_type_id: int | None = None,
        character_id: int | None = None,
        corporation_id: int | None = None,
        alliance_id: int | None = None,
        npc: bool | None = None,
        solo: bool | None = None,
        awox: bool | None = None,
    ) -> tuple[list[dict], int]:
        clauses = []
        params: dict = {}

        if min_value is not None:
            clauses.append("k.zkb_total_value >= :min_value")
            params["min_value"] = min_value
        if max_value is not None:
            clauses.append("k.zkb_total_value <= :max_value")
            params["max_value"] = max_value
        if solar_system_id is not None:
            clauses.append("k.solar_system_id = :solar_system_id")
            params["solar_system_id"] = solar_system_id
        if ship_type_id is not None:
            clauses.append("k.victim_ship_type_id = :ship_type_id")
            params["ship_type_id"] = ship_type_id
        if character_id is not None:
            clauses.append(
                "(k.victim_character_id = :character_id"
                " OR EXISTS (SELECT 1 FROM killmail_attackers a"
                " WHERE a.killmail_id = k.killmail_id"
                " AND a.character_id = :character_id))"
            )
            params["character_id"] = character_id
        if corporation_id is not None:
            clauses.append(
                "(k.victim_corporation_id = :corporation_id"
                " OR EXISTS (SELECT 1 FROM killmail_attackers a"
                " WHERE a.killmail_id = k.killmail_id"
                " AND a.corporation_id = :corporation_id))"
            )
            params["corporation_id"] = corporation_id
        if alliance_id is not None:
            clauses.append(
                "(k.victim_alliance_id = :alliance_id"
                " OR EXISTS (SELECT 1 FROM killmail_attackers a"
                " WHERE a.killmail_id = k.killmail_id"
                " AND a.alliance_id = :alliance_id))"
            )
            params["alliance_id"] = alliance_id
        if npc is not None:
            clauses.append("k.zkb_is_npc = :npc")
            params["npc"] = int(npc)
        if solo is not None:
            clauses.append("k.zkb_is_solo = :solo")
            params["solo"] = int(solo)
        if awox is not None:
            clauses.append("k.zkb_is_awox = :awox")
            params["awox"] = int(awox)

        where = " AND ".join(clauses) if clauses else "1=1"

        total = self.session.execute(
            text(f"SELECT COUNT(*) FROM killmails k WHERE {where}"), params
        ).scalar()

        rows = self.session.execute(
            text(
                f"SELECT k.* FROM killmails k WHERE {where}"
                " ORDER BY k.killmail_time DESC LIMIT :limit OFFSET :offset"
            ),
            {**params, "limit": limit, "offset": offset},
        )

        kills = []
        for r in rows:
            row = dict(r._mapping)
            row["zkb_labels"] = json.loads(row["zkb_labels"]) if row.get("zkb_labels") else []
            kills.append(row)

        return kills, total

    def get_kill(self, killmail_id: int) -> dict | None:
        row = self.session.execute(
            text("SELECT * FROM killmails WHERE killmail_id = :id"),
            {"id": killmail_id},
        ).first()

        if row is None:
            return None

        kill = dict(row._mapping)
        kill["zkb_labels"] = json.loads(kill["zkb_labels"]) if kill.get("zkb_labels") else []

        attackers = self.session.execute(
            text("SELECT * FROM killmail_attackers WHERE killmail_id = :id ORDER BY damage_done DESC"),
            {"id": killmail_id},
        )
        kill["attackers"] = [dict(r._mapping) for r in attackers]

        items = self.session.execute(
            text("SELECT * FROM killmail_items WHERE killmail_id = :id"),
            {"id": killmail_id},
        )
        kill["items"] = self._build_item_tree([dict(r._mapping) for r in items])

        return kill

    def get_stats(self) -> dict:
        row = self.session.execute(
            text(
                "SELECT COUNT(*) AS total_kills, COALESCE(SUM(zkb_total_value), 0) AS total_value,"
                " SUM(zkb_is_npc) AS kills_npc, SUM(zkb_is_solo) AS kills_solo,"
                " SUM(zkb_is_awox) AS kills_awox FROM killmails"
            )
        ).first()

        stats = dict(row._mapping)

        ships = self.session.execute(
            text(
                "SELECT victim_ship_type_id AS ship_type_id, COUNT(*) AS count"
                " FROM killmails GROUP BY victim_ship_type_id ORDER BY count DESC LIMIT 10"
            )
        )
        stats["top_ships"] = [dict(r._mapping) for r in ships]

        systems = self.session.execute(
            text(
                "SELECT solar_system_id, COUNT(*) AS count"
                " FROM killmails GROUP BY solar_system_id ORDER BY count DESC LIMIT 10"
            )
        )
        stats["top_solar_systems"] = [dict(r._mapping) for r in systems]

        return stats

    @staticmethod
    def _build_item_tree(flat_items: list[dict]) -> list[dict]:
        by_id = {item["id"]: {**item, "items": []} for item in flat_items}
        roots = []
        for item_id, item in by_id.items():
            parent = item.pop("parent_id", None)
            if parent is None or parent not in by_id:
                roots.append(item)
            else:
                by_id[parent]["items"].append(item)
        return roots

File: typescript/src/repository.ts
import type { Pool, PoolConnection, RowDataPacket, ResultSetHeader } from "mysql2/promise";
import type {
  R2Z2Killmail,
  KillmailSummary,
  KillmailDetail,
  AttackerRow,
  ItemNode,
  Stats,
  KillListQuery,
} from "./types.js";

export class KillmailRepository {
  constructor(private pool: Pool) {}

  // ── Write (poller) ──────────────────────────────────────────────

  async save(killmail: R2Z2Killmail): Promise<boolean> {
    const conn = await this.pool.getConnection();
    try {
      await conn.beginTransaction();

      const { esi, zkb } = killmail;
      const { victim } = esi;
      const pos = victim.position;

      const [result] = await conn.execute<ResultSetHeader>(
        `INSERT IGNORE INTO killmails (
          killmail_id, hash, killmail_time, solar_system_id, sequence_id,
          war_id, moon_id,
          victim_character_id, victim_corporation_id, victim_alliance_id,
          victim_faction_id, victim_ship_type_id, victim_damage_taken,
          victim_pos_x, victim_pos_y, victim_pos_z,
          zkb_location_id, zkb_fitted_value, zkb_dropped_value,
          zkb_destroyed_value, zkb_total_value, zkb_points,
          zkb_is_npc, zkb_is_solo, zkb_is_awox,
          zkb_labels, zkb_href, zkb_attacker_count, uploaded_at
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
        [
          killmail.killmail_id,
          killmail.hash,
          esi.killmail_time,
          esi.solar_system_id,
          killmail.sequence_id,
          esi.war_id ?? null,
          esi.moon_id ?? null,
          victim.character_id ?? null,
          victim.corporation_id ?? null,
          victim.alliance_id ?? null,
          victim.faction_id ?? null,
          victim.ship_type_id,
          victim.damage_taken,
          pos?.x ?? null,
          pos?.y ?? null,
          pos?.z ?? null,
          zkb.locationID ?? null,
          zkb.fittedValue ?? 0,
          zkb.droppedValue ?? 0,
          zkb.destroyedValue ?? 0,
          zkb.totalValue ?? 0,
          zkb.points ?? 0,
          zkb.npc ? 1 : 0,
          zkb.solo ? 1 : 0,
          zkb.awox ? 1 : 0,
          JSON.stringify(zkb.labels ?? []),
          zkb.href ?? null,
          (esi.attackers ?? []).length,
          new Date(killmail.uploaded_at * 1000).toISOString().slice(0, 19).replace("T", " "),
        ]
      );

      if (result.affectedRows === 0) {
        await conn.rollback();
        return false;
      }

      await this.insertAttackers(conn, killmail.killmail_id, esi.attackers ?? []);
      await this.insertItems(conn, killmail.killmail_id, victim.items ?? []);

      await conn.commit();
      return true;
    } catch (e) {
      await conn.rollback();
      throw e;
    } finally {
      conn.release();
    }
  }

  private async insertAttackers(
    conn: PoolConnection,
    killmailId: number,
    attackers: R2Z2Killmail["esi"]["attackers"] & {}
  ): Promise<void> {
    for (const a of attackers) {
      await conn.execute(
        `INSERT INTO killmail_attackers (
          killmail_id, character_id, corporation_id, alliance_id, faction_id,
          ship_type_id, weapon_type_id, damage_done, final_blow, security_status
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
        [
          killmailId,
          a.character_id ?? null,
          a.corporation_id ?? null,
          a.alliance_id ?? null,
          a.faction_id ?? null,
          a.ship_type_id ?? null,
          a.weapon_type_id ?? null,
          a.damage_done ?? 0,
          a.final_blow ? 1 : 0,
          a.security_status ?? 0,
        ]
      );
    }
  }

  private async insertItems(
    conn: PoolConnection,
    killmailId: number,
    items: R2Z2Killmail["esi"]["victim"]["items"] & {},
    parentId: number | null = null
  ): Promise<void> {
    for (const item of items) {
      const [result] = await conn.execute<ResultSetHeader>(
        `INSERT INTO killmail_items (
          killmail_id, parent_id, item_type_id, flag,
          quantity_destroyed, quantity_dropped, singleton
        ) VALUES (?, ?, ?, ?, ?, ?, ?)`,
        [
          killmailId,
          parentId,
          item.item_type_id,
          item.flag ?? 0,
          item.quantity_destroyed ?? 0,
          item.quantity_dropped ?? 0,
          item.singleton ?? 0,
        ]
      );

      if (item.items?.length) {
        await this.insertItems(conn, killmailId, item.items, result.insertId);
      }
    }
  }

  // ── Read (API) ──────────────────────────────────────────────────

  async listKills(query: KillListQuery): Promise<{ kills: KillmailSummary[]; total: number }> {
    const clauses: string[] = [];
    const params: unknown[] = [];

    if (query.min_value != null) {
      clauses.push("k.zkb_total_value >= ?");
      params.push(query.min_value);
    }
    if (query.max_value != null) {
      clauses.push("k.zkb_total_value <= ?");
      params.push(query.max_value);
    }
    if (query.solar_system_id != null) {
      clauses.push("k.solar_system_id = ?");
      params.push(query.solar_system_id);
    }
    if (query.ship_type_id != null) {
      clauses.push("k.victim_ship_type_id = ?");
      params.push(query.ship_type_id);
    }
    if (query.character_id != null) {
      clauses.push(
        "(k.victim_character_id = ? OR EXISTS (SELECT 1 FROM killmail_attackers a WHERE a.killmail_id = k.killmail_id AND a.character_id = ?))"
      );
      params.push(query.character_id, query.character_id);
    }
    if (query.corporation_id != null) {
      clauses.push(
        "(k.victim_corporation_id = ? OR EXISTS (SELECT 1 FROM killmail_attackers a WHERE a.killmail_id = k.killmail_id AND a.corporation_id = ?))"
      );
      params.push(query.corporation_id, query.corporation_id);
    }
    if (query.alliance_id != null) {
      clauses.push(
        "(k.victim_alliance_id = ? OR EXISTS (SELECT 1 FROM killmail_attackers a WHERE a.killmail_id = k.killmail_id AND a.alliance_id = ?))"
      );
      params.push(query.alliance_id, query.alliance_id);
    }
    if (query.npc != null) {
      clauses.push("k.zkb_is_npc = ?");
      params.push(query.npc ? 1 : 0);
    }
    if (query.solo != null) {
      clauses.push("k.zkb_is_solo = ?");
      params.push(query.solo ? 1 : 0);
    }
    if (query.awox != null) {
      clauses.push("k.zkb_is_awox = ?");
      params.push(query.awox ? 1 : 0);
    }

    const where = clauses.length ? clauses.join(" AND ") : "1=1";
    const limit = Math.min(query.limit ?? 50, 1000);
    const offset = query.offset ?? 0;

    const [countRows] = await this.pool.execute<RowDataPacket[]>(
      `SELECT COUNT(*) AS total FROM killmails k WHERE ${where}`,
      params
    );
    const total = countRows[0].total as number;

    const [rows] = await this.pool.execute<RowDataPacket[]>(
      `SELECT k.killmail_id, k.killmail_time, k.solar_system_id,
        k.victim_ship_type_id, k.victim_character_id, k.victim_corporation_id,
        k.victim_alliance_id, k.zkb_total_value, k.zkb_is_npc, k.zkb_is_solo,
        k.zkb_is_awox, k.zkb_attacker_count
      FROM killmails k WHERE ${where}
      ORDER BY k.killmail_time DESC LIMIT ? OFFSET ?`,
      [...params, limit, offset]
    );

    const kills: KillmailSummary[] = rows.map((r) => ({
      ...r,
      zkb_is_npc: !!r.zkb_is_npc,
      zkb_is_solo: !!r.zkb_is_solo,
      zkb_is_awox: !!r.zkb_is_awox,
    })) as KillmailSummary[];

    return { kills, total };
  }

  async getKill(killmailId: number): Promise<KillmailDetail | null> {
    const [rows] = await this.pool.execute<RowDataPacket[]>(
      "SELECT * FROM killmails WHERE killmail_id = ?",
      [killmailId]
    );

    if (rows.length === 0) return null;

    const row = rows[0];
    row.zkb_labels = JSON.parse(row.zkb_labels ?? "[]");
    row.zkb_is_npc = !!row.zkb_is_npc;
    row.zkb_is_solo = !!row.zkb_is_solo;
    row.zkb_is_awox = !!row.zkb_is_awox;

    const [attackerRows] = await this.pool.execute<RowDataPacket[]>(
      "SELECT * FROM killmail_attackers WHERE killmail_id = ? ORDER BY damage_done DESC",
      [killmailId]
    );
    const attackers: AttackerRow[] = attackerRows.map((a) => ({
      ...a,
      final_blow: !!a.final_blow,
    })) as AttackerRow[];

    const [itemRows] = await this.pool.execute<RowDataPacket[]>(
      "SELECT * FROM killmail_items WHERE killmail_id = ?",
      [killmailId]
    );
    const items = this.buildItemTree(
      itemRows.map((i) => ({ ...i, singleton: !!i.singleton })) as (RowDataPacket & { parent_id: number | null })[]
    );

    return { ...row, attackers, items } as KillmailDetail;
  }

  async getStats(): Promise<Stats> {
    const [summaryRows] = await this.pool.execute<RowDataPacket[]>(
      `SELECT COUNT(*) AS total_kills, COALESCE(SUM(zkb_total_value), 0) AS total_value,
        SUM(zkb_is_npc) AS kills_npc, SUM(zkb_is_solo) AS kills_solo,
        SUM(zkb_is_awox) AS kills_awox FROM killmails`
    );
    const s = summaryRows[0];

    const [shipRows] = await this.pool.execute<RowDataPacket[]>(
      `SELECT victim_ship_type_id AS ship_type_id, COUNT(*) AS count
       FROM killmails GROUP BY victim_ship_type_id ORDER BY count DESC LIMIT 10`
    );

    const [systemRows] = await this.pool.execute<RowDataPacket[]>(
      `SELECT solar_system_id, COUNT(*) AS count
       FROM killmails GROUP BY solar_system_id ORDER BY count DESC LIMIT 10`
    );

    return {
      total_kills: s.total_kills,
      total_value: Number(s.total_value),
      kills_npc: Number(s.kills_npc ?? 0),
      kills_solo: Number(s.kills_solo ?? 0),
      kills_awox: Number(s.kills_awox ?? 0),
      top_ships: shipRows as Stats["top_ships"],
      top_solar_systems: systemRows as Stats["top_solar_systems"],
    };
  }

  private buildItemTree(flatItems: (RowDataPacket & { parent_id: number | null })[]): ItemNode[] {
    const byId = new Map<number, ItemNode>();

    for (const item of flatItems) {
      byId.set(item.id as number, {
        id: item.id as number,
        item_type_id: item.item_type_id as number,
        flag: item.flag as number,
        quantity_destroyed: item.quantity_destroyed as number,
        quantity_dropped: item.quantity_dropped as number,
        singleton: !!item.singleton,
        items: [],
      });
    }

    const roots: ItemNode[] = [];
    for (const item of flatItems) {
      const node = byId.get(item.id as number)!;
      if (item.parent_id == null || !byId.has(item.parent_id)) {
        roots.push(node);
      } else {
        byId.get(item.parent_id)!.items.push(node);
      }
    }

    return roots;
  }
}
